### PR TITLE
LinearAlgebra: diagzero for non-OneTo axes

### PR DESCRIPTION
**Pull request by jishnub**
Currently, the off-diagonal zeros for a block-`Diagonal` matrix is computed using `diagzero`, which calls `zeros` for the sizes of the elements. This returns an `Array`, unless one specializes `diagzero` for the custom `Diagonal` matrix type.

This PR defines a `zeroslike` function that dispatches on the axes of the elements, which lets packages specialize on the axes to return custom `AbstractArray`s. Choosing to specialize on the `eltype` avoids the need to specialize on the container, and allows packages to return appropriate types for custom axis types.

With this,
```julia
julia> LinearAlgebra.zeroslike(::Type{S}, ax::Tuple{SOneTo, Vararg{SOneTo}}) where {S<:SMatrix} = SMatrix{map(length, ax)...}(ntuple(_->zero(eltype(S)), prod(length, ax)))

julia> D = Diagonal(fill(SMatrix{2,3}(1:6), 2))
2×2 Diagonal{SMatrix{2, 3, Int64, 6}, Vector{SMatrix{2, 3, Int64, 6}}}:
 [1 3 5; 2 4 6]        ⋅       
       ⋅         [1 3 5; 2 4 6]

julia> D[1,2] # now an SMatrix
2×3 SMatrix{2, 3, Int64, 6} with indices SOneTo(2)×SOneTo(3):
 0  0  0
 0  0  0

julia> LinearAlgebra.zeroslike(::Type{S}, ax::Tuple{SOneTo, Vararg{SOneTo}}) where {S<:MMatrix} = MMatrix{map(length, ax)...}(ntuple(_->zero(eltype(S)), prod(length, ax)))

julia> D = Diagonal(fill(MMatrix{2,3}(1:6), 2))
2×2 Diagonal{MMatrix{2, 3, Int64, 6}, Vector{MMatrix{2, 3, Int64, 6}}}:
 [1 3 5; 2 4 6]        ⋅       
       ⋅         [1 3 5; 2 4 6]

julia> D[1,2] # now an MMatrix
2×3 MMatrix{2, 3, Int64, 6} with indices SOneTo(2)×SOneTo(3):
 0  0  0
 0  0  0
```
The reason this can't be the default behavior is that we are not guaranteed that there exists a `similar` method that accepts the combination of axes. This is why we have to fall back to using the sizes, unless a specialized method is provided by a package.

One positive outcome of this is that indexing into such a block-diagonal matrix will now usually be type-stable, which mitigates https://github.com/JuliaLang/LinearAlgebra.jl/issues/927 to some extent (although it doesn't resolve the issue).

I've also updated the `getindex` for `Bidiagonal` to use `diagzero`, instead of the similarly defined `bidiagzero` function that it was using. Structured block matrices may now use `diagzero` uniformly to generate the zero elements.

### CODE REVIEW

**Comment by mikmoore** [comment]
Can the `LinearAlgebra.AbstractTriangular` family also be covered by this PR? In fact, they appear to not even be correct in some cases:
```julia-repl
julia> UpperTriangular([ones(i,j) for i in 1:2, j in 1:2])
2×2 UpperTriangular{Matrix{Float64}, Matrix{Matrix{Float64}}}:
 [1.0;;]    [1.0 1.0]
     ⋅      [1.0 1.0; 1.0 1.0]

julia> UpperTriangular([ones(i,j) for i in 1:2, j in 1:2])[1,2]
1×2 Matrix{Float64}:
 1.0  1.0

julia> UpperTriangular([ones(i,j) for i in 1:2, j in 1:2])[2,1] # output has wrong size!
1×2 Matrix{Float64}:
 0.0  0.0
```

**Comment by jishnub** [comment]
Sounds reasonable to me, but that may be a separate PR

File: stdlib/LinearAlgebra/src/diagonal.jl [hunk 1]
```diff
@@ -185,8 +185,28 @@ end
...
 end
-diagzero(::Diagonal{T}, i, j) where {T} = zero(T)
-diagzero(D::Diagonal{<:AbstractMatrix{T}}, i, j) where {T} = zeros(T, size(D.diag[i], 1), size(D.diag[j], 2))
+"""
+    diagzero(A::AbstractMatrix, i, j)
+
+Return the appropriate zero element `A[i, j]` corresponding to a banded matrix `A`.
+"""
+diagzero(A::AbstractMatrix, i, j) = zero(eltype(A))
+diagzero(D::Diagonal{M}, i, j) where {T,M<:AbstractMatrix{T}} =
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/diagonal.jl:194]:
```suggestion
diagzero(D::Diagonal{M}, i, j) where {M<:AbstractMatrix} =
```

You don't seem to use `T` anywhere?

File: stdlib/LinearAlgebra/src/diagonal.jl [hunk 2]
```diff
@@ -185,8 +185,28 @@ end
...
+
+Return an appropriate zero-ed array similar to `M`, with either
+the axes `ax`, or the `size` `map(length, ax)`.
+This will be used as a structural zero element of a banded matrix. By default, `zeroslike` falls back to
+using the size along each axis to construct the array.
+"""
+zeroslike(M::Type, ax::Tuple{AbstractUnitRange, Vararg{AbstractUnitRange}}) = zeroslike(M, map(length, ax))
+function zeroslike(::Type{M}, sz::Tuple{Integer, Vararg{Integer}}) where {M}
+    zeros(M <: AbstractMatrix ? eltype(M) : M, sz)
+end
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/diagonal.jl:209]: Can't you write this with multiple dispatch? Like

```julia
zeroslike(::Type{M}, sz::Tuple{Integer, Vararg{Integer}}) where {M <: AbstractMatrix} = zeros(eltype(M), sz)
zeroslike(::Type{M}, sz::Tuple{Integer, Vararg{Integer}}) where {M} = zeros(M, sz)
```

**Review by dkarrasch — COMMENTED** [review]
LGTM. Should this mechanism be communicated to package maintainers in NEWS.md?

**Pull request merged**